### PR TITLE
Implement QuadraticBezierCurve and MANNAutoregressiveInputBuilder

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This PR implements the `MANNAutoregressiveInputBuilder` that is required to easily generate input for MANN from a joypad. The class does not depend on the joypad since it requires only two vectors. The desired motion direction and the facing direction.

### CODE REVIEW

File: src/ML/src/MANNAutoregressiveInputBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+    if (m_pimpl->input.motionDirection[0] >= 0)
+    {
+        m_pimpl->curve.controlPoint[0]
+            = m_pimpl->ellipsoidScalingFactor
+              * std::min(m_pimpl->ellipsoidForwardAxis, m_pimpl->input.motionDirection[0]);
+        m_pimpl->curve.finalPoint[0] = m_pimpl->input.motionDirection[0] * m_pimpl->ellipsoidScalingFactor
+                        * m_pimpl->ellipsoidForwardAxis;
+        m_pimpl->curve.finalPoint[1] = m_pimpl->input.motionDirection[1] * m_pimpl->ellipsoidScalingFactor
+                        * m_pimpl->ellipsoidSideAxis;
```
Comment (S-Dafarra) [inline src/ML/src/MANNAutoregressiveInputBuilder.cpp:226]: This part seems the same on both cases. You may want to move it outside the if for readability sake

File: src/ML/src/MANNAutoregressiveInputBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    m_pimpl->output.desiredFutureBaseTrajectory
+        = m_pimpl->curve.curve.evaluateCurve(m_pimpl->curve.initialPoint, m_pimpl->curve.controlPoint, m_pimpl->curve.finalPoint);
+
+    m_pimpl->output.desiredFutureBaseVelocities.rightCols(m_pimpl->numberOfKnots - 1)
+        = m_pimpl->output.desiredFutureBaseTrajectory.rightCols(m_pimpl->numberOfKnots - 1)
+          - m_pimpl->output.desiredFutureBaseTrajectory.leftCols(m_pimpl->numberOfKnots - 1);
```
Comment (S-Dafarra) [inline src/ML/src/MANNAutoregressiveInputBuilder.cpp:243]: Why is this necessary?

File: src/ML/src/MANNAutoregressiveInputBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+double MANNAutoregressiveInputBuilder::Impl::evaluateMaxFacingDirectionAngle(
+    Eigen::Ref<const Eigen::Vector2d> normalizedMotionDirection,
+    Eigen::Ref<const Eigen::Vector2d> rawFacingDirection) const
+{
+    if (normalizedMotionDirection[0] < 0.2)
+    {
+        return maxFacingDirectionAngleBackward;
+    }
+
+    if (std::abs(normalizedMotionDirection[1]) > 0.2)
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp:66]: I agree @paolo-viceconte do you have any suggestion for the name of these parameters?

File: src/ML/src/MANNAutoregressiveInputBuilder.cpp (hunk 2 above)
Comment (GiulioRomualdi) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp]: Here I added some comments in the code that explain the rationale behind https://github.com/ami-iit/bipedal-locomotion-framework/pull/683/commits/e335c02a26e55760544e2d69ea1b686e753d4c7d

moreover 
```cpp
    m_pimpl->output.desiredFutureBaseVelocities.rightCols(m_pimpl->numberOfKnots - 1)
        = m_pimpl->output.desiredFutureBaseTrajectory.rightCols(m_pimpl->numberOfKnots - 1)
          - m_pimpl->output.desiredFutureBaseTrajectory.leftCols(m_pimpl->numberOfKnots - 1);
```
should become
```cpp
    m_pimpl->output.desiredFutureBaseVelocities.leftCols(m_pimpl->numberOfKnots - 1)
        = m_pimpl->output.desiredFutureBaseTrajectory.rightCols(m_pimpl->numberOfKnots - 1)
          - m_pimpl->output.desiredFutureBaseTrajectory.leftCols(m_pimpl->numberOfKnots - 1);
```

File: src/ML/src/MANNAutoregressiveInputBuilder.cpp (hunk 1 above)
Comment (GiulioRomualdi) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp]: Done in https://github.com/ami-iit/bipedal-locomotion-framework/commit/e335c02a26e55760544e2d69ea1b686e753d4c7d

File: src/ML/src/MANNAutoregressiveInputBuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        m_pimpl->output.desiredFutureBaseTrajectory.resize(2, m_pimpl->numberOfKnots);
+        m_pimpl->output.desiredFutureBaseVelocities.resize(2, m_pimpl->numberOfKnots);
+        m_pimpl->output.desiredFutureFacingDirections.resize(2, m_pimpl->numberOfKnots);
+
+        m_pimpl->curve.initialPoint.setZero();
+        m_pimpl->frontalDirection << 1, 0;
+
+        m_pimpl->fsm = Impl::FSM::Initialized;
+    }
+
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp:204]: Added in https://github.com/ami-iit/bipedal-locomotion-framework/commit/e335c02a26e55760544e2d69ea1b686e753d4c7d

File: src/ML/include/BipedalLocomotion/ML/MANNAutoregressiveInputBuilder.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+     * |                 `base_vel_norm`                 | `double` |                                              Norm of the desired base velocity in m/s                                               |    Yes    |
+     * |             `ellipsoid_forward_axis`            | `double` |                                             Size of the forward axis ellipsoid in meters                                            |    Yes    |
+     * |            `ellipsoid_backward_axis`            | `double` |                                            Size of the backward axis ellipsoid in meters                                            |    Yes    |
+     * |              `ellipsoid_side_axis`              | `double` |                                             Size of the lateral axis ellipsoid in meters                                            |    Yes    |
+     * |            `ellipsoid_scaling_factor`           | `double` |                    Scaling factor considered in the generation of the final and control points o the Bezier Curve                   |    Yes    |
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply src/ML/include/BipedalLocomotion/ML/MANNAutoregressiveInputBuilder.h]: Thank you! Fixed in https://github.com/ami-iit/bipedal-locomotion-framework/pull/683/commits/7a96e1a53dfc395bdc6314d061d4e66fa76ff4cd

File: bindings/python/ML/include/BipedalLocomotion/bindings/ML/MANNAutoregressiveInputBuilder.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
+/*
+ * @file MANNAutoregressiveInputBuilder.h
+ * @authors Giulio Romualdi
+ * @copyright 2023 Istituto Italiano di Tecnologia (IIT). This software may be modified and
+ * distributed under the terms of the BSD-3-Clause license.
+ */
+
+#ifndef BIPEDAL_LOCOMOTION_BINDINGS_ML_MANN_AUTOREGRESSIVE_TRAJECTORY_BUILDER_H
+#define BIPEDAL_LOCOMOTION_BINDINGS_ML_MANN_AUTOREGRESSIVE_TRAJECTORY_BUILDER_H
```
[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply bindings/python/ML/include/BipedalLocomotion/bindings/ML/MANNAutoregressiveInputBuilder.h]: Ops! Fixed in https://github.com/ami-iit/bipedal-locomotion-framework/pull/683/commits/7427ba99eae31bd17cf2e93a8ca9ffbd3bb4881b

File: src/ML/src/MANNAutoregressiveInputBuilder.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (paolo-viceconte) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp:65]: We want to differentiate here the kind of motion commanded by the user by looking at the user-specified motion direction. To do so, we portion the space of motion directions (normalized in [0,1]) into different regions. Those magic numbers represent the borders of the various regions. What about `minForwardMotionDirection` for the first 0.2 (line 61) and `minSideMotionDirection` for the second 0.2 (line 66)?

[the remark this answers is not in the document]
Comment (S-Dafarra) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp:65]: I see. In this case I would suggest the ``threshold`` keyword. Maybe ``forwardDirectionThreshold`` and ``sideDirectionThreshold``?

[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp:65]: > I see. In this case I would suggest the `threshold` keyword. Maybe `forwardDirectionThreshold` and `sideDirectionThreshold`?

Deal! I'm going to add these two parameters as optional with a default value of 0.2

[the remark this answers is not in the document]
Comment (GiulioRomualdi) [reply src/ML/src/MANNAutoregressiveInputBuilder.cpp]: Done in https://github.com/ami-iit/bipedal-locomotion-framework/pull/683/commits/d3d5b3cc0a6349b9a0a804ef2ea534946649f257

**Comment by GiulioRomualdi** [comment]
I applied the modification you suggested, @S-Dafarra you can check again it 😄

**Review by S-Dafarra — APPROVED** [review]
